README.md
# bitmask

bitmask is a Dart package that provides a Bitmask class.

## Introduction

A Bitmask object contains
a collection of individual bits that may be set or not set (1 or 0).
Constructors and methods are provided for setting and unsetting individual
bits in a Bitmask and for applying operations such as bitwise ANDing, ORing, and XORing masks together, taking a bitwise complement, and for
getting the
integer value represented by the Bitmask object.

Bitmask objects are limited
to representing 63 bits as this is the largest number of bits that can be
contained in a Dart integer. If this class is used on the Web, the maximum
number of bits should be limited to 52 as this is the largest number of bits
that can be represented reliably by Javascript.

Here is a simple example showing how to use the Bitmask class with
enumerations:

```dart
import 'package:bitmask/bitmask.dart';

enum Maskbits { zero, one, two, three, four, }

void main() {
  var mask1 = Bitmask.fromList([Maskbits.one.index, 
      Maskbits.three.index],
      Maskbits.values().length);
  mask1[Maskbits.four] = true;
  var mask2 = Bitmask.fromBitmask(mask1);
  mask2.unset(Maskbits.three.index);
  print(mask1.flags.toRadixString(2)); // prints 11010
  print((mask1 & mask2).flags.toRadixString(2)); // prints 10010
  print((mask1 | mask2).flags.toRadixString(2)); // prints 11010
  print((mask1 ^ mask2).flags.toRadixString(2)); // prints 1000
  print(~mask1.flags.toRadixString(2)); // prints 101
}
```

## Installation

Add `bitmask` to you `pubspec.yaml` file.

```yaml
dependencies:
  bitmask: ^1.0.0
```

After saving the `pubspec.yaml` file, run:

```bash
dart pub update
```

## Contributions

If you would like to contribute to the bitmask library, please see
[CONTRIBUTING.md](https://github.com/jimorc/bitmask/blob/main/CONTRIBUTING.md).

CHANGELOG.md
## 1.0.1

- Update description in pubspec.yaml. pub.dev complained it was too short.

## 1.0.0

- Initial version.


test/bitmask_test.dart
import 'package:bitmask/bitmask.dart';
import 'package:test/test.dart';

enum MaskBits {
  zero,
  one,
  two,
  three,
}

void main() {
  group('Bitmask tests', () {
    test('empty constructor', () {
      var mask = Bitmask(MaskBits.values.length);
      expect(mask.flags, 0);
    });

    (test('invalid Bitmask sizes', () {
      expect(() => Bitmask(-1), throwsA(isArgumentError));
      expect(() => Bitmask(0), throwsA(isArgumentError));
      expect(Bitmask(1).flags, 0);
      expect(Bitmask(63).flags, 0);
      expect(() => Bitmask(64).flags, throwsA(isArgumentError));
    }));

    test('fromInt', () {
      var bits = 13;
      var mask = Bitmask.fromInt(bits, 6);
      expect(mask.flags, bits);
    });

    test('fromInt invalid size', () {
      expect(() => Bitmask.fromInt(13, 0), throwsA(isArgumentError));
      expect(() => Bitmask.fromInt(13, -1), throwsA(isArgumentError));
      expect(() => Bitmask.fromInt(13, 64), throwsA(isArgumentError));
      expect(Bitmask.fromInt(13, 1).flags, 1);
      expect(Bitmask.fromInt(13, 63).flags, 13);
    });

    test('set a bit', () {
      var mask = Bitmask(MaskBits.values.length);
      mask.set(MaskBits.two.index);
      mask.set(MaskBits.three.index);
      expect(mask.flags, 12);
    });

    test('set invalid bit', () {
      var mask = Bitmask(MaskBits.values.length);
      expect(() => mask.set(-1), throwsA(isArgumentError));
      mask.set(0);
      mask.set(3);
      expect(() => mask.set(4), throwsA(isArgumentError));
    });

    test('unset a bit', () {
      var mask = Bitmask(MaskBits.values.length);
      mask.set(MaskBits.two.index);
      mask.set(MaskBits.three.index);
      expect(mask.flags, 12);
      mask.unset(MaskBits.two.index);
      expect(mask.flags, 8);
    });

    test('unset invalid bit', () {
      var mask = Bitmask(MaskBits.values.length);
      expect(() => mask.unset(-1), throwsA(isArgumentError));
      mask.unset(0);
      mask.unset(3);
      expect(() => mask.unset(4), throwsA(isArgumentError));
    });

    test('isSet', () {
      var mask = Bitmask(MaskBits.values.length);
      mask.set(MaskBits.one.index);
      mask.set(MaskBits.three.index);
      expect(mask.isSet(MaskBits.zero.index), false);
      expect(mask.isSet(MaskBits.one.index), true);
      expect(mask.isSet(MaskBits.two.index), false);
      expect(mask.isSet(MaskBits.three.index), true);
    });

    test('invalid bit in isSet', () {
      var mask = Bitmask(MaskBits.values.length);
      expect(() => mask.isSet(-1), throwsA(isArgumentError));
      var _ = mask.isSet(0);
      expect(() => mask.isSet(4), throwsA(isArgumentError));
    });

    test('operator []', () {
      var mask = Bitmask(MaskBits.values.length);
      mask.set(MaskBits.one.index);
      mask.set(MaskBits.three.index);
      expect(mask[MaskBits.zero.index], false);
      expect(mask[MaskBits.one.index], true);
      expect(mask[MaskBits.two.index], false);
      expect(mask[MaskBits.three.index], true);
    });

    test('invalid bit in []', () {
      var mask = Bitmask(MaskBits.values.length);
      expect(() => mask[-1], throwsA(isArgumentError));
      mask.set(0);
      expect(mask.flags, 1);
      expect(() => mask[4], throwsA(isArgumentError));
    });

    test('operator []= to set a bit', () {
      var mask = Bitmask(MaskBits.values.length);
      mask[MaskBits.one.index] = true;
      mask[MaskBits.three.index] = true;
      expect(mask.isSet(MaskBits.zero.index), false);
      expect(mask.isSet(MaskBits.one.index), true);
      expect(mask.isSet(MaskBits.two.index), false);
      expect(mask.isSet(MaskBits.three.index), true);

      mask[MaskBits.two.index] = false;
      mask[MaskBits.three.index] = false;
      expect(mask.isSet(MaskBits.zero.index), false);
      expect(mask.isSet(MaskBits.one.index), true);
      expect(mask.isSet(MaskBits.two.index), false);
      expect(mask.isSet(MaskBits.three.index), false);
    });

    test('invalid bit in operator []=', () {
      var mask = Bitmask(MaskBits.values.length);
      expect(() => mask[-1] = true, throwsA(isArgumentError));
      var _ = mask.isSet(0);
      expect(() => mask[4] = true, throwsA(isArgumentError));
    });

    test('operator ~', () {
      var mask = Bitmask(MaskBits.values.length);
      mask[MaskBits.one.index] = true;
      mask[MaskBits.three.index] = true;
      var mask2 = ~mask;
      expect(mask2.flags, 5);
      expect(identical(mask, mask2), false);
    });

    test('bitwise AND of two Bitmasks', () {
      var mask = Bitmask(MaskBits.values.length);
      mask[MaskBits.one.index] = true;
      mask[MaskBits.three.index] = true;
      var mask2 = Bitmask(MaskBits.values.length);
      mask2[MaskBits.one.index] = true;
      expect((mask & mask2).flags, 2);
      expect(identical(mask, mask2), false);
    });

    test('bitwise AND of two Bitmasks that are not same size', () {
      var mask = Bitmask(4);
      var mask2 = Bitmask(6);
      expect(() => mask & mask2, throwsA(isArgumentError));
    });

    test('bitwise OR of two Bitmasks', () {
      var mask = Bitmask(MaskBits.values.length);
      mask[MaskBits.one.index] = true;
      mask[MaskBits.three.index] = true;
      var mask2 = Bitmask(MaskBits.values.length);
      mask2[MaskBits.two.index] = true;
      expect((mask | mask2).flags, 14);
      expect(identical(mask, mask2), false);
    });

    test('bitwise OR of two Bitmasks that are not same size', () {
      var mask = Bitmask(4);
      var mask2 = Bitmask(6);
      expect(() => mask | mask2, throwsA(isArgumentError));
    });

    test('bitwise XOR of two Bitmasks', () {
      var mask = Bitmask(MaskBits.values.length);
      mask[MaskBits.one.index] = true;
      mask[MaskBits.three.index] = true;
      var mask2 = Bitmask(MaskBits.values.length);
      mask2[MaskBits.two.index] = true;
      mask2[MaskBits.one.index] = true;
      expect((mask ^ mask2).flags, 12);
      expect(identical(mask, mask2), false);
    });

    test('bitwise XOR of two Bitmasks that are not same size', () {
      var mask = Bitmask(4);
      var mask2 = Bitmask(6);
      expect(() => mask ^ mask2, throwsA(isArgumentError));
    });

    test('operator ==', () {
      var mask = Bitmask(4);
      expect(mask == mask, true);

      var object = Object();
      expect(mask == object, false);

      var mask2 = Bitmask(4);
      expect(mask == mask2, true);

      mask[1] = true;
      expect(mask == mask2, false);

      var mask3 = Bitmask(6);
      expect(mask == mask3, false);

      mask2[2] = true;
      expect(mask == mask2, false);
    });

    test('operator !=', () {
      var mask = Bitmask(4);
      expect(mask != mask, false);

      var object = Object();
      expect(mask != object, true);

      var mask2 = Bitmask(4);
      expect(mask != mask2, false);

      mask[1] = true;
      expect(mask != mask2, true);

      var mask3 = Bitmask(6);
      expect(mask != mask3, true);

      mask2[2] = true;
      expect(mask != mask2, true);
    });

    test('fromList', () {
      var mask = Bitmask.fromList([1, 3, 5], 8);
      expect(mask.flags, 42);
    });

    test('invalid input in fromList', () {
      // test various mask sizes
      expect(() => Bitmask.fromList([], -1), throwsA(isArgumentError));
      expect(() => Bitmask.fromList([], 0), throwsA(isArgumentError));
      var _ = Bitmask.fromList([], 1);
      _ = Bitmask.fromList([], 63);
      expect(() => Bitmask.fromList([], 64), throwsA(isArgumentError));

      // test invalid bits in list
      expect(() => Bitmask.fromList([4, -1, 5], 6), throwsA(isArgumentError));
      expect(() => Bitmask.fromList([4, 1, 6], 6), throwsA(isArgumentError));
      _ = Bitmask.fromList([0, 5], 6);
    });

    test('fromBitmask', () {
      var mask = Bitmask.fromList([1, 3, 5], 6);
      var mask2 = Bitmask.fromBitmask(mask);
      expect(mask2 == mask, true);
      expect(identical(mask, mask2), false);
      mask.unset(3);
      expect(mask != mask2, true);
      expect(identical(mask, mask2), false);
    });

    test('clear', () {
      var mask = Bitmask.fromList([1, 3, 5], 6);
      expect(mask.flags, 42);
      mask.clear();
      expect(mask.flags, 0);
    });

    test('hashcode', () {
      var mask = Bitmask.fromList([1, 3, 5], 6);
      expect(mask.flags, 42);
      expect(mask.hashCode, 42.hashCode);
    });
  });
}


example/bitmask_example.dart
import 'package:bitmask/bitmask.dart';

enum Maskbits { zero, one, two, three, four }

void main() {
  var mask1 = Bitmask(6);
  // set bits 1 and 3.
  mask1[1] = true;
  mask1.set(3); // same as mask1[3] = true;
  // set, then unset bit 5
  mask1[5] = true;
  mask1.unset(5); // same as mask1[5] = false;

  // sets bits 0, 2, and 3.
  var mask2 = Bitmask.fromInt(13, 6); // sets bits 0, 2, and 3.

  // set bits 0, 3, and 4.
  var mask3 = Bitmask.fromList(
      [Maskbits.zero.index, Maskbits.three.index, Maskbits.four.index], 6);

  print('mask1.flags = ${mask1.flags.toRadixString(2)}');
  print('mask2.flags = ${mask2.flags.toRadixString(2)}');
  print('mask3.flags = ${mask3.flags.toRadixString(2)}');

  print('mask1 & mask2 = ${(mask1 & mask2).flags.toRadixString(2)}');
  print('mask1 & mask3 = ${(mask1 & mask3).flags.toRadixString(2)}');
  print('mask2 & mask3 = ${(mask2 & mask3).flags.toRadixString(2)}');

  print('mask1 | mask2 = ${(mask1 | mask2).flags.toRadixString(2)}');
  print('mask1 | mask3 = ${(mask1 | mask3).flags.toRadixString(2)}');
  print('mask2 | mask3 = ${(mask2 | mask3).flags.toRadixString(2)}');

  print('~mask1 = ${(~mask1).flags.toRadixString(2)}');
  print('~mask2 = ${(~mask2).flags.toRadixString(2)}');
  print('~mask3 = ${(~mask3).flags.toRadixString(2)}');

  var mask4 = Bitmask.fromBitmask(mask3);
  print('mask4 = ${mask4.flags.toRadixString(2)}');
  print('mask4 == mask3: ${mask4 == mask3}');
  print('identical(mask4, mask3): ${identical(mask4, mask3)}');

  mask3.clear();
  print('After mask3.clear(), mask3 = ${mask3.flags.toRadixString(2)}');

  print('mask4 bit 4: ${mask4.isSet(4)}');
  print('mask4 bit 2: ${mask4[2]}');
}

// prints the following:
// mask1.flags = 1010
// mask2.flags = 1101
// // mask3.flags = 11001
// mask1 & mask2 = 1000
// mask1 & mask3 = 1000
// mask2 & mask3 = 1001
// mask1 | mask2 = 1111
// mask1 | mask3 = 11011
// mask2 | mask3 = 11101
// ~mask1 = 110101
// ~mask2 = 110010
// ~mask3 = 100110
// mask4 = 11001
// mask4 == mask3: true
// identical(mask4, mask3): false
// After mask3.clear(), mask3 = 0
// mask4 bit 4: true
// mask4 bit 2: false

lib/src/bitmask_base.dart
const int _minimumBitmaskSize = 1;
const int _maximumBitmaskSize = 63;

/// A Bitmask class, which contains a collection of single bits.
///
/// There are four ways to construct a Bitmap object:
/// ```dash
/// // Create a Bitmask of 6 bits all set to 0
/// final mask1 = Bitmask(6);
/// // Create a Bitmask of 10 bits, with bits 1, 3, and 5 set to 1.
/// final mask2 = Bitmask.fromInt(21, 10);
/// // Create a Bitmask of 4 bits, with bits 1 and 3 set.
/// final mask3 = Bitmask.fromList([1, 3], 4);
/// // Create a Bitmask that is a deep copy of another.
/// final mask4 = Bitmask.fromBitmask(mask2);
/// ```
/// There are two ways to set individual bits, and two ways to unset
/// individual bits:
/// ```dash
/// // To set a bit, use either:
/// mask2.set(4); // Set bit 4 to 1 (true);
/// mask2[4] = true;  // Set the same bit to 1 (true)
/// // To unset a bit, use either:
/// mask2.unset(3);  // Set bit 3 to 0 (false)
/// mask2[3] = false; // Set same bit to 0 (false)
/// ```
/// The normal bit-wise operations are supported:
/// ```dash
/// // Bit-wise AND.
/// var mask5 = mask1 & mask2;
/// // Bit-wise OR.
/// var mask6 = mask1 | mask2;
/// // Bit-wise XOR.
/// var mask7 = mask1 ^ mask2;
/// // Bit-wise NOT.
/// var mask8 = ~mask1;
///```
///To convert to and from an integer:
///```dash
/// // Create an 8 bit Bitmask object with bits 1, 3, and 5 set.
/// var mask9 = Bitmask.fromInt(21, 8);
/// // Get integer value from Bitmask object
/// var bits = mask9.flags; // bits contains value 21.
/// ```
/// To reset all bits in a Bitmask object to 0 (false), call clear().
/// ```dash
/// mask9.clear();  // mask9 now contains all zeroes.
/// ```
/// There are two ways to determine if a bit is set:
/// ```dart
/// bool bit4 = mask7.isSet(4);
/// bool bit3 = mask7[3];
/// ```
/// Dart does not support generic enumerations, so Bitmask constructors
/// and methods use integer arguments. It is possible to use enumerations
/// with Bitmask by converting the emumeration values to integers as follows:
/// ```dash
/// enum Maskbits { zero, one, two, three, four }
///
/// var mask1 = Bitmask(Maskbits.values().length);
/// mask1[Maskbits.one.index] = true; // Set bit 1.
/// // Create Bitmask object to hold all Maskbits, with bits 1 and 2 set to 1
/// var mask2 = Bitmask.fromList([Maskbits.one.index, Maskbits.two.index],
///     Maskbits.values().length);
/// ```
/// Here is a program that exercises the Bitmask class. It is the same
/// as the example program:
/// ```dash
/// import 'package:bitmask/bitmask.dart';
///
/// enum Maskbits { zero, one, two, three, four }
///
/// void main() {
///   var mask1 = Bitmask(6);
///   // set bits 1 and 3.
///   mask1[1] = true;
///   mask1.set(3); // same as mask1[3] = true;
///   // set, then unset bit 5
///   mask1[5] = true;
///   mask1.unset(5); // same as mask1[5] = false;
///
///   // sets bits 0, 2, and 3.
///   var mask2 = Bitmask.fromInt(13, 6); // sets bits 0, 2, and 3.
///
///   // set bits 0, 3, and 4.
///   var mask3 = Bitmask.fromList(
///       [Maskbits.zero.index, Maskbits.three.index, Maskbits.four.index], 6);
///
///   print('mask1.flags = ${mask1.flags.toRadixString(2)}');
///   print('mask2.flags = ${mask2.flags.toRadixString(2)}');
///   print('mask3.flags = ${mask3.flags.toRadixString(2)}');
///
///   print('mask1 & mask2 = ${(mask1 & mask2).flags.toRadixString(2)}');
///   print('mask1 & mask3 = ${(mask1 & mask3).flags.toRadixString(2)}');
///   print('mask2 & mask3 = ${(mask2 & mask3).flags.toRadixString(2)}');
///
///   print('mask1 | mask2 = ${(mask1 | mask2).flags.toRadixString(2)}');
///   print('mask1 | mask3 = ${(mask1 | mask3).flags.toRadixString(2)}');
///   print('mask2 | mask3 = ${(mask2 | mask3).flags.toRadixString(2)}');
///
///   print('~mask1 = ${(~mask1).flags.toRadixString(2)}');
///   print('~mask2 = ${(~mask2).flags.toRadixString(2)}');
///   print('~mask3 = ${(~mask3).flags.toRadixString(2)}');
///
///   var mask4 = Bitmask.fromBitmask(mask3);
///   print('mask4 = ${mask4.flags.toRadixString(2)}');
///   print('mask4 == mask3: ${mask4 == mask3}');
///   print('identical(mask4, mask3): ${identical(mask4, mask3)}');
///
///   mask3.clear();
///   print('After mask3.clear(), mask3 = ${mask3.flags.toRadixString(2)}');
///
///   print('mask4 bit 4: ${mask4.isSet(4)}');
///   print('mask4 bit 2: ${mask4[2]}');
/// }
///
/// // prints the following:
/// // mask1.flags = 1010
/// // mask2.flags = 1101
/// // mask3.flags = 11001
/// // mask1 & mask2 = 1000
/// // mask1 & mask3 = 1000
/// // mask2 & mask3 = 1001
/// // mask1 | mask2 = 1111
/// // mask1 | mask3 = 11011
/// // mask2 | mask3 = 11101
/// // ~mask1 = 110101
/// // ~mask2 = 110010
/// // ~mask3 = 100110
/// // mask4 = 11001
/// // mask4 == mask3: true
/// // identical(mask4, mask3): false
/// // After mask3.clear(), mask3 = 0
/// // mask4 bit 4: true
/// // mask4 bit 2: false
/// ```
class Bitmask {
  /// Construct a bitmask with no bits set.
  ///
  /// _numberOfBits_ is the number of bits in the BitMask. The minimum
  /// value for this argument is 1; a Bitmask of zero or a negative
  /// size does not make sense. The maximum
  /// value for this argument is 63 as that is the size of a bitmask that
  /// can fit in an integer. This value should be no larger than 52 if
  /// the BitMask is used on the Web because that is the largest size that
  /// can be reliably set in Javascript.
  ///
  /// Throws ArgumentError if the bit number is invalid (< 1 or greater
  /// than 63).
  Bitmask(int numberOfBits) {
    if (numberOfBits < _minimumBitmaskSize) {
      throw ArgumentError('Bitmask constructor: Attempting to create a '
          "Bitmask size less than one bit.");
    }
    if (numberOfBits > _maximumBitmaskSize) {
      throw ArgumentError('Bitmask constructor: Attempting to create a '
          'Bitmask size larger than 63 bits in size.');
    }
    _mask = List<bool>.filled(numberOfBits, false, growable: false);
  }

  /// Create a Bitmask from an integer value.
  ///
  /// _bits_ is the integer value to create the Bitmask for.
  ///
  /// _numberOfBits_ is the number of bits in the Bitmask. No check is made
  /// to ensure that the size of the Bitmask is large enough to hold all
  /// of the bits in _bits_. The minimum
  /// value for this argument is 1; a Bitmask of zero or a negative
  /// size does not make sense. The maximum
  /// value for this argument is 63 as that is the size of a bitmask that
  /// can fit in an integer. This value should be no larger than 52 if
  /// the BitMask is used on the Web because that is the largest size that
  /// can be reliably set in Javascript.
  ///
  /// throws ArgumentError is _size_ is not between 1 and 63.
  factory Bitmask.fromInt(int bits, int numberOfBits) {
    if (numberOfBits < _minimumBitmaskSize ||
        numberOfBits > _maximumBitmaskSize) {
      throw ArgumentError('Bitmask.fromInt: Bitmask size must be between'
          ' 1 and 63.');
    }
    Bitmask mask = Bitmask(numberOfBits);
    for (int i = 0; i < numberOfBits; i++) {
      mask[i] = bits & 1 == 1;
      bits >>= 1;
    }
    return mask;
  }

  /// Create Bitmask with bits specified in a list set.
  ///
  /// _bits_ is the list of bits to set.
  ///
  /// _numberOfBits_ is the number of bits in this Bitmask.
  ///
  /// Throws ArgumentError if _size_ is not between 1 and 63.
  ///
  /// Throws ArgumentError if any entry in _bits_ is outside the range
  /// of 0 and _size_ - 1.
  factory Bitmask.fromList(List<int> bits, int numberOfBits) {
    if (numberOfBits < _minimumBitmaskSize ||
        numberOfBits > _maximumBitmaskSize) {
      throw ArgumentError('Bitmask.fromList: Bitmask size must be between'
          ' 1 and 63.');
    }
    var mask = Bitmask(numberOfBits);
    for (var bit in bits) {
      if (bit < _minimumBitmaskSize - 1 || bit >= numberOfBits) {
        throw ArgumentError('Bitmask.fromList: Attempting to set a bit '
            'that is less than 0 or greater than the size of the Bitmask.');
      }
      mask[bit] = true;
    }
    return mask;
  }

  /// Create a Bitmask object that is a copy of another Bitmask object.
  ///
  /// This constructor is required because with
  /// ```dart
  /// var mask2 = mask1;
  /// ```
  /// both _mask1_ and _mask2_ point to the same object, not different objects
  /// with the same content. Also, dart does
  /// not allow overriding _operator =_, so we can't override it to
  /// produce the result that we want.
  factory Bitmask.fromBitmask(Bitmask other) {
    return Bitmask.fromInt(other.flags, other.length);
  }

  /// Set all bits to false.
  void clear() {
    _mask = List<bool>.filled(_mask.length, false, growable: false);
  }

  /// Get the bitmask as an integer.
  int get flags {
    var result = 0;
    for (var bit = 0; bit < _mask.length; bit++) {
      if (_mask[bit]) {
        result += 1 << bit;
      }
    }
    return result;
  }

  /// Set the bit specified by _bit_ to 1 (true).
  ///
  /// Throws ArgumentError if the bit number is invalid (< 0 or greater
  /// than the number of bits in the mask).
  void set(int bit) {
    if (bit < _minimumBitmaskSize - 1 || bit > _mask.length - 1) {
      throw ArgumentError('BitMask.set: Request to set bit \'$bit\' which'
          'is invalid.');
    }
    _mask[bit] = true;
  }

  // Set the bit specified by _bit_ to 0 (false).
  ///
  /// Throws ArgumentError if the bit number is invalid (< 0 or greater
  /// than the number of bits in the mask).
  void unset(int bit) {
    if (bit < _minimumBitmaskSize - 1 || bit > _mask.length - 1) {
      throw ArgumentError('BitMask.set: Request to set bit \'$bit\' which'
          'is invalid.');
    }
    _mask[bit] = false;
  }

  /// Check if a bit is set.
  ///
  /// Returns true if the bit is set, and false if the bit is not set.
  ///
  /// Throws ArgumentError if the bit number is invalid (< 0 or greater
  /// than the number of bits in the mask).
  bool isSet(int bit) {
    if (bit < _minimumBitmaskSize - 1 || bit > _mask.length - 1) {
      throw ArgumentError('Bitmask.isSet: Request to check bit: \'$bit\''
          ' is invalid.');
    }
    return _mask[bit];
  }

  /// Get the setting of a bit.
  ///
  /// Returns true if the bit is set, and false if the bit is not set.
  /// than the number of bits in the mask).
  ///
  /// Throws ArgumentError if the bit number is invalid (< 0 or greater
  /// than the number of bits in the mask).
  bool operator [](int bit) => isSet(bit);

  /// Set a bit to 1 (true) or 0 (false).
  ///
  /// Throws ArgumentError if the bit number is invalid (< 0 or greater
  /// than the number of bits in the mask).
  operator []=(int bit, bool value) {
    if (bit < _minimumBitmaskSize - 1 || bit >= _mask.length) {
      throw ArgumentError('Bitmask.operator []=: Request to set bit: '
          '\'$bit\' is invalid.');
    }
    _mask[bit] = value;
  }

  /// Create a Bitmask of the bitwise complement of this object.
  Bitmask operator ~() {
    var newMask = Bitmask.fromBitmask(this);
    for (var bit = 0; bit < _mask.length; bit++) {
      newMask[bit] = !_mask[bit];
    }
    return newMask;
  }

  /// Create a Bitmask that is the bitwise AND of one Bitmask with another.
  ///
  /// Throws ArgumentError if the Bitmasks are not the same size.
  Bitmask operator &(Bitmask other) {
    if (length != other.length) {
      throw ArgumentError('Bitmask.operator &: Attempting to do a bitwise'
          ' AND of two Bitmasks that are not the same size.');
    }
    Bitmask newMask = Bitmask.fromBitmask(this);
    for (int bit = 0; bit < _mask.length; bit++) {
      newMask[bit] &= other[bit];
    }
    return newMask;
  }

  /// Create a Bitmask that is the bitwise OR of one Bitmask with another.
  ///
  /// Throws ArgumentError if the Bitmasks are not the same size.
  Bitmask operator |(Bitmask other) {
    if (length != other.length) {
      throw ArgumentError('Bitmask.operator |: Attempting to do a bitwise'
          ' AND of two Bitmasks that are not the same size.');
    }
    Bitmask newMask = Bitmask.fromBitmask(this);
    for (var bit = 0; bit < _mask.length; bit++) {
      newMask[bit] |= other[bit];
    }
    return newMask;
  }

  /// Create a Bitmask that is the bitwise XOR of one Bitmask with another.
  ///
  /// Throws ArgumentError if the Bitmasks are not the same size.
  Bitmask operator ^(Bitmask other) {
    if (length != other.length) {
      throw ArgumentError('Bitmask.operator |: Attempting to do a bitwise'
          ' AND of two Bitmasks that are not the same size.');
    }
    Bitmask newMask = Bitmask.fromBitmask(this);
    for (var bit = 0; bit < length; bit++) {
      newMask[bit] ^= other[bit];
    }
    return newMask;
  }

  /// Hash code of this object.
  @override
  int get hashCode => flags.hashCode;

  /// Equality operator.
  ///
  /// For two Bitmasks to be equal, they must be the same size and have
  /// all of the same bits set and unset.
  @override
  bool operator ==(Object other) {
    return other is Bitmask && length == other.length && flags == other.flags;
  }

  /// The number of bits in the mask.
  int get length => _mask.length;

  late List<bool> _mask;
}
